### DISCUSSION TITLE
Clarification on Top Level Imports

### DISCUSSION THREAD

**Discussion by pwu-380** [apache/airflow#26177] [Q&A] [answered]
Hello,

I think we have some Airflow performance issues at my company that may be related to the following:
https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#best-practices-top-level-code 

But I'm having a hard time interpreting what "top level code" means in the context of our codebase. We have hundreds of deployed dags written by many different business units, but here are a couple examples I've seen (simplified for clarity):

## Scenario 1:
my_dag.py
```
import airflow
from airflow.operators import operator1, operator2, PythonOperator

def main():
    dag_args = {
        ...
    }

    create_dag(dag_args)

def create_dag(dag_args):
    from mynamespace.mymodule import helper1, helper2

    with airflow.DAG(**dag_args) as dag:
        task1 = operator1(...)
       
        task2 = operator2(...)
       
        task3 = PythonOperator(python_callable=helper1)
       
        task4 = PythonOperator(python_callable=helper2)

        (task1 >> task2 >> task3 >> task4)

    globals()[dag.dag_id] = dag

main()
```
mymodule.py
```
import biglibrary1
import biglibrary2

def helper1(...):
    ...

def helper2(...):
    ...
```
Will biglibrary1 and biglibrary2 be a problem? Should we be putting their imports into helper1 and helper2?

## Scenario 2:
mydag.py
```
import airflow

def main():
    config = {
        ...
    }

    create_dag(config)

def create_dag(config):
    from mynamespace.mymodule import DagFactory

    dag_factory = DagFactory(config)
    globals()[dag_factory.dag_id] = dag_factory.get_dag()

main()
```
mymodule.py
The DagFactory might be doing things, e.g. creating new cloud resources/checking if certain resources exist and then creating Airflow tasks that reference them
```
import airflow
import biglibrary1
import biglibrary2
from mynamespace.myothermodule import Things
from mynamespace.myotherothermodule import MoreThings
from airflow.operators import operator1, operator2

class DagFactory():
    def __init__(...):
        ...
        self.dag = airflow.DAG(**dag_args)
        ...
        self._chain_tasks(list_of_tasks)

    def _method1(...):
        ...
   
    def _method2(...):
        ...

    def _taskmethod1(...):
        ...
        new_task = operator1(dag=self.dag)
        return new_task

    def _taskmethod2(...):
        ...
        new_task = operator2(dag=self.dag)
        return new_task

    def _chain_tasks(list_of_tasks: List[operators]):
        # This method is defining the order for all the tasks
        ...
       
    def get_dag(...):
        return self.dag
```
Now I'm really losing grip on what's "top level" here. Is the prescription to encapsulate all imports into the method that uses them? Is using a factory an anti-pattern to begin with? 

Thanks for any feedback.

**Comment by potiuk** [comment]
Top level is everything that gets executed when python code is parsed. This is nicely explained in this SO answer: https://stackoverflow.com/questions/24356801/how-does-the-python-executable-parse-and-execute-scripts#:~:text=As%20a%20general%20rule%2C%20python,time%2C%20before%20anything%20is%20executed. and boils down to this statement:

```
 Every statement in your module, except those contained within def or lambda, is executed in order. 
```
^^ This is the code that is not top-level code

**Comment by potiuk** [comment]
One caveat, that might be not obvious though. Your `main()` function from first example also gets executed - so what is inside `def main()` gets executed as well.  Even if it is inside of "def" - because you explicity executed it via:

```
main()
```

Which instructs parser to execute what has been defined in:

```
def main():
      ....
```

**Comment by potiuk** [comment, marked as answer]
And more detailed answers:

> Will biglibrary1 and biglibrary2 be a problem? Should we be putting their imports into helper1 and helper2?

Yes - it will be a problem. Parsing Python code is literally executing every line of it (except those inside def and lambdas unless explicitly executed elsewhere). Executing "import x" - loads and executes `__init__.py` of the `x` module. You can avoid that by making your imports local inside `def` that you do not execute (putting it in `main` will still import it (main gets executed), but inside `helper1` will skip the intialization and it will be called only when helper1 is called. 

> # Scenario 2 
> Now I'm really losing grip on what's "top level" here. Is the prescription to encapsulate all imports into the method that uses them? Is using a factory an anti-pattern to begin with?


In your case you call `main()` explicitly when the code is parsed. The main calls `create_dag()` and `create_dag` calls both `DagFactory().__init__` (when creating the object) and as next step it calls `get_dag()` of the created DagFactory object. All of the code get executed during parsing.

Simply speaking - moving any code to inside the `def ` function  and then calling the `def` function explicitly in top level code (which the code you copied seems to be doing) - does not work.  The `main()` at the end of your dag will get it executed.

**Reply by pwu-380** [reply]
Hi Potiuk, thank you for taking the time to explain (and being so quick!)

> And more detailed answers:
> 
> > Will biglibrary1 and biglibrary2 be a problem? Should we be putting their imports into helper1 and helper2?
> 
> Yes - it will be a problem. Parsing Python code is literally executing every line of it (except those inside def and lambdas unless explicitly executed elsewhere). Executing "import x" - loads and executes `__init__.py` of the `x` module. You can avoid that by making your imports local inside `def` that you do not execute (putting it in `main` will still import it (main gets executed), but inside `helper1` will skip the intialization and it will be called only when helper1 is called.
> 
> > # Scenario 2
> > Now I'm really losing grip on what's "top level" here. Is the prescription to encapsulate all imports into the method that uses them? Is using a factory an anti-pattern to begin with?
> 
> In your case you call `main()` explicitly when the code is parsed. The main calls `create_dag()` and `create_dag` calls both `DagFactory().__init__` (when creating the object) and as next step it calls `get_dag()` of the created DagFactory object. All of the code get executed during parsing.
> 
> Simply speaking - moving any code to inside the `def ` function and then calling the `def` function explicitly in top level code (which the code you copied seems to be doing) - does not work. The `main()` at the end of your dag will get it executed.

So let me rephrase to check my understanding, any Python code that gets executed will have to be parsed on a regular interval from the Airflow scheduler. In the Airflow doc example, and scenario 1, encapsulating imports into a function helps because `print_array()`/`helper1()`/`helper2()`  is not called anywhere from main, only an alias is passed as a parameter to an operator. Therefore Airflow won't parse their encapsulated code until runtime. Is that the gist?

**Reply by potiuk** [reply]
It's not called from anywhere when parsing. It does not have to be main. The probkem is that you ARE running main() in your DAG. This is the root cause of the problem and something that should not happen in general, because it is literally hiding the "top-level code" under whatever "main()" method does.

**Reply by potiuk** [reply]
Again -  this is pure python,. Run it via `python dag.py` and you will see what it does, you can add print() statement and you will see it and it will be much easier to grasp. 

Simply accept the fact that Python interpreter parsing the Python file EXECUTES it line-by-line - with the exception of `def` and `lambdas` which define a "deferred" code. This `deferred` code is only executed when someone calls the function/method/lambda defined. So if one of the lines that parser executes is `main()` this means that at this moment, the parser will execute the code that was previously skipped under `def main:`. And it's recursive.

**Comment by potiuk** [comment]
BTW. This just pure python behaviour - nothing "Airflow specific". So you can very easily see what gets executed during parsing by running:

```
python your_dag.py
```

There is no "airflow magic" here at all.
